contracts/Registry.sol

```diff
@@ -133,6 +210,58 @@ contract Registry is IRegistry {
         );
     }
 
+    /**
+     * @dev Alter a DSNP Id handle by EIP-712 Signature
+     * @param v EIP-155 calculated Signature v value
+     * @param r ECDSA Signature r value
+     * @param s ECDSA Signature s value
+     * @param change Change data containing nonce, old handle and new handle
+     */
+    function changeHandleByEIP712Sig(
+        uint8 v,
+        bytes32 r,
+        bytes32 s,
+        HandleChange calldata change
+    ) external override {
+        // Checks
+
+        Registration storage oldReg = registrations[change.oldHandle];
+        require(oldReg.id != 0, "Old handle does not exist");
+        require(oldReg.nonce == change.nonce, "Nonces do not match");
+
+        Registration storage newReg = registrations[change.newHandle];
+        require(newReg.id == 0, "New handle already exists");
+
+        address signer = handleChangeSigner(v, r, s, change);
+
+        // Effects
+
+        // assign to new registration
+        newReg.id = oldReg.id;
+        newReg.identityAddress = oldReg.identityAddress;
+
+        // signal that the old handle is unassigned and available
+        oldReg.id = 0;
+
+        // increment nonce so this transaction cannot be replayed
+        oldReg.nonce++;
+
+        // notify the change
+        emit DSNPRegistryUpdate(newReg.id, newReg.identityAddress, change.newHandle);
+
+        // Interactions
+
+        IDelegation authorization = IDelegation(oldReg.identityAddress);
```

Comment:
again do we want to check for permissions before we emit the Registry update event

Reply:
This is the checks, effects, interactions pattern. We want all interactions with other contracts to occur after all updates to avoid reentrancy issues. This probably isn't strictly necessary for events since they can't be viewed from other contracts, but the event won't be emitted if the contract is reverted, so it also doesn't hurt to emit it before checking authorization and I kind of like the consistency of putting it with the other updates.

Reply:
oh that totally makes sense. Thanks!